File: clients/logger/src/main/clojure/samsara/slf4j/SamsaraLoggerFactory.clj
(ns samsara.slf4j.SamsaraLoggerFactory
  (:import [samsara.slf4j SamsaraLogger]
           [org.slf4j Logger])
  (:gen-class :implements [org.slf4j.ILoggerFactory]
              :constructors {[] []}
              :state state
              :init init))


(defn- -init []
  [[] (atom {:logger-map {}})]) ; initialise logger-map with an atom

(defn- -getLogger ^Logger[this ^String name]
  (locking this
    (if-let [logger (get-in @(.state this) [:logger-map name])]
      logger
      (let [logger (SamsaraLogger. name)]
        (swap! (.state this) assoc-in [:logger-map  name] logger)
        logger))))

File: clients/logger/src/main/java/samsara/slf4j/SamsaraLoggerFactory.java
package samsara.slf4j;

import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.ConcurrentHashMap;

import org.slf4j.Logger;
import org.slf4j.ILoggerFactory;

public class SamsaraLoggerFactory implements ILoggerFactory
{
    private ConcurrentMap<String, Logger> loggerMap = new ConcurrentHashMap<>();

    public Logger getLogger(String name)
    {
        Logger samsaraLogger = loggerMap.get(name);
        if(samsaraLogger != null)
        {
            return samsaraLogger;
        }
        else
        {
            Logger newLogger = new SamsaraLogger(name);
            Logger oldLogger = loggerMap.putIfAbsent(name, newLogger);
            return oldLogger == null ? newLogger : oldLogger;
        }
    }
}

